### PR TITLE
fix optional wait wrongly treated as false

### LINKED ISSUE #77984: When dummy argument "wait" of intrinsic "execute_command_line" is not present, it's treated as ".false."
Here's a program that shows the problem:
```
program bug
  call inner("echo hello")
contains
  subroutine inner(cmd, wait, exitstatarg, cmdstatarg, cmdmsgarg)
    character(10) :: cmd
    logical, optional :: wait
    integer, optional :: exitstatarg, cmdstatarg
    character(128), optional :: cmdmsgarg
    integer :: localvar
    call execute_command_line(cmd, wait, exitstatarg, localvar, cmdmsgarg)
  end subroutine
end
```

@yi-wu-arm, please take note.
When I compile and executes this program, it behaves as if the `wait` argument has the value `.false.`.  It's supposed to act as if the value is `.true.`.

### PR DESCRIPTION
**Pull request by yiwu0b11**
When an optional bool in passed as optional, `isStaticallyPresent()` will actually return true.
By changing wait from `asValue` to `asAddr` and add some check can solve this.
Fixes: https://github.com/llvm/llvm-project/issues/77984

### CODE REVIEW

**Comment by yi-wu-arm** [comment]
When testing with
```fortran
program bug
  call inner("sleep 2 && echo hi")
  print *, "Direct printout line"
contains
  subroutine inner(cmd, wait, exitstatarg, cmdstatarg, cmdmsgarg)
    character(20) :: cmd
    logical, optional :: wait
    integer, optional :: exitstatarg, cmdstatarg
    character(128), optional :: cmdmsgarg
    integer :: localvar
    call execute_command_line(cmd, wait, exitstatarg, localvar, cmdmsgarg)
  end subroutine
end
```
The console print out would be:
```
```
And when set wait to false,
```
```

File: flang/test/Lower/Intrinsics/execute_command_line.f90 [hunk 1]
```diff
@@ -39,7 +46,7 @@ end subroutine all_args
 subroutine only_command_default_wait_true(command)
 CHARACTER(30) :: command
 call execute_command_line(command)
-! CHECK-NEXT:     %c41_i32 = arith.constant 41 : i32 
+! CHECK-NEXT:     %c48_i32 = arith.constant 48 : i32
```
Comment (kiranchandramohan) [inline flang/test/Lower/Intrinsics/execute_command_line.f90:49]: Using the name of the constant might be prone to error. Capture this in a value and use that value for checking.

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 2]
```diff
@@ -2913,7 +2913,8 @@ IntrinsicLibrary::genEoshift(mlir::Type resultType,
 // EXECUTE_COMMAND_LINE
 void IntrinsicLibrary::genExecuteCommandLine(
     llvm::ArrayRef<fir::ExtendedValue> args) {
-  assert(args.size() == 5);
+  // Optional KIND argument.
+  assert(args.size() >= 1);
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:2917]: There is no KIND argument for `execute_command_line`. Is there a typo?

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp [hunk 3]
```diff
@@ -2925,9 +2926,30 @@ void IntrinsicLibrary::genExecuteCommandLine(
...
+                     mlir::Value cast =
+                         builder.createConvert(loc, i1Ty, waitLoad);
+                     builder.create<fir::ResultOp>(loc, cast);
+                   })
+                   .genElse([&]() {
+                     mlir::Value trueVal = builder.createBool(loc, true);
+                     builder.create<fir::ResultOp>(loc, trueVal);
+                   })
+                   .getResults()[0];
+  }
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/Builder/IntrinsicCall.cpp:2951]: Is there similar handling for any other intrinsic?

File: flang/test/Lower/Intrinsics/execute_command_line.f90 (hunk 1 above)
Comment (yi-wu-arm) [reply flang/test/Lower/Intrinsics/execute_command_line.f90]: done

**Review by psteinfeld — COMMENTED** [review]
All builds and tests correctly and looks good to me.

You might want to wait for Kiran's approval before merging.

File: flang/lib/Optimizer/Builder/IntrinsicCall.cpp (hunk 3 above)
Comment (luporl) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp:2952]: https://reviews.llvm.org/D123388 explains the logic behind this code.

Comment (kiranchandramohan) [reply flang/lib/Optimizer/Builder/IntrinsicCall.cpp:2952]: Thanks for the pointers @yi-wu-arm @luporl

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Pull request merged**